core/executor/src/native_executor.rs

```diff
@@ -59,28 +62,78 @@ pub trait NativeExecutionDispatch: Send + Sync {
 	fn native_version() -> NativeVersion;
 }
 
+/// The callable Substrate runtime. When called, a version of the Wasm runtime code is executed.
+pub trait SubstrateRuntime {
+	/// Call a method in the Substrate runtime by name. Returns the encoded result on success.
+	fn call(&mut self, ext: &mut dyn Externalities<Blake2Hasher>, method: &str, data: &[u8])
+		-> Result<Vec<u8>>;
+
+	/// Returns the version of this runtime.
+	///
+	/// Returns `None` if the runtime doesn't provide the information or there was an error
+	/// while fetching it.
+	fn version(&self) -> Option<RuntimeVersion>;
+}
+
+/// Specification of different methods of executing the runtime Wasm code.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum WasmExecutionMethod {
+	/// Uses the Wasmi interpreter.
+	Interpreted,
+}
+
+impl Default for WasmExecutionMethod {
+	fn default() -> Self {
+		WasmExecutionMethod::Interpreted
+	}
+}
+
 /// A generic `CodeExecutor` implementation that uses a delegate to determine wasm code equivalence
 /// and dispatch to native code when possible, falling back on `WasmExecutor` when not.
 #[derive(Debug)]
 pub struct NativeExecutor<D> {
 	/// Dummy field to avoid the compiler complaining about us not using `D`.
 	_dummy: ::std::marker::PhantomData<D>,
-	/// The fallback executor in case native isn't available.
-	fallback: WasmExecutor,
+	/// Method used to execute fallback Wasm code.
+	wasm_method: WasmExecutionMethod,
 	/// Native runtime version info.
 	native_version: NativeVersion,
 	/// The number of 64KB pages to allocate for Wasm execution.
-	default_heap_pages: Option<u64>,
+	default_heap_pages: u64,
 }
 
 impl<D: NativeExecutionDispatch> NativeExecutor<D> {
 	/// Create new instance.
-	pub fn new(default_heap_pages: Option<u64>) -> Self {
+	///
+	/// # Parameters
+	///
+	/// `wasm_method` - Method used to execute fallback Wasm code.
+	///
+	/// `default_heap_pages` - Number of 64KB pages to allocate for Wasm execution.
+	/// Defaults to `DEFAULT_HEAP_PAGES` if `None` is provided.
+	pub fn new(wasm_method: WasmExecutionMethod, default_heap_pages: Option<u64>) -> Self {
 		NativeExecutor {
 			_dummy: Default::default(),
-			fallback: WasmExecutor::new(),
+			wasm_method,
 			native_version: D::native_version(),
-			default_heap_pages: default_heap_pages,
+			default_heap_pages: default_heap_pages.unwrap_or(DEFAULT_HEAP_PAGES),
+		}
+	}
+
+	fn with_runtime<E, R>(
+		&self,
+		ext: &mut E,
+		f: impl for <'a> FnOnce(Box<dyn SubstrateRuntime + 'a>, &'a mut E) -> Result<R>
+	) -> Result<R>
+		where E: Externalities<Blake2Hasher>
+	{
+		match self.wasm_method {
+			WasmExecutionMethod::Interpreted =>
+				RUNTIMES_CACHE.with(|cache| {
+					let mut cache = cache.borrow_mut();
+					let runtime = cache.fetch_runtime(&WasmiExecutor, ext, self.default_heap_pages)?;
+					f(Box::new(runtime), ext)
```

Comment:
Why do you box the value here?

Reply:
If `self.wasm_method` is different, it might be a different impl of SubstrateRuntime, so it needs to be a trait object. Would you prefer it be a `&'a dyn SubstrateRuntime`?

Reply:
Yes. Box requires allocating memory.